### ISSUE TITLE
EFI Stub: Support for hashed external UKI components

### ISSUE THREAD

**Issue by ElvishJerricco** [systemd/systemd#26096]
### Component

systemd-stub

### Is your feature request related to a problem? Please describe

For distributions that store many initrds and kernels in the boot partition, UKIs present a problem. You need a somewhat large file for each of the combination of components that you want to have. For instance, there are ongoing experimental efforts to support secure boot with NixOS, and it's not uncommon to have dozens, or even hundreds, of cmdline alternatives because each profile generation gets a new cmdline.

### Describe the solution you'd like

It would be nice if the things stored in the UKI's PE sections could be optionally stored on the file system instead, and the UKI could include the expected hashes of those components. The stub would read these components from the file system instead and verify the hash. Because the hash is in the UKI, it would be verified by secure boot. For NixOS's case, we could have one very small UKI per generation, and have them point to their shared kernels/initrds on the file system instead of copying them into yet another UKI.

### Describe alternatives you've considered

https://github.com/systemd/systemd/issues/24539 is almost certainly the easier and better solution to the NixOS example. However, I can still imagine reasonably wanting to be able to mix and match many UKI components and reduce duplication in this way. For instance, if you had five initrds for one kernel and wanted a boot option for each of them, you could reduce the space usage by the equivalent of four kernels by separating it into an external file.

### The systemd version you checked that didn't have the feature you are asking for

252

**Comment by bluca** [comment]
If you have to rebuild the UKI to change the hash then surely you can just rebuild the UKI to include the update component? The main advantage of the Unified Kernel Image is that, well, it's unified...

**Comment by poettering** [comment]
initrds usually contain kmods and stuff, thus they are actually pretty tightly bound to specific kernels. Thus such a "mix & match" doesn't really make much sense, does it?

**Comment by ElvishJerricco** [comment]
I may not have made it clear; this is about space usage. @bluca it isn't about the process of rebuilding the UKI. Having a hundred full UKIs would take a lot of space on what is usually a small partition. @poettering yes, you can't really pair a different kernel with the same initrd, but you can pair many initrds with the same kernel, and that does in fact happen sometimes in NixOS. As I said:

> For instance, if you had five initrds for one kernel and wanted a boot option for each of them, you could reduce the space usage by the equivalent of four kernels by separating it into an external file.

**Comment by bluca** [comment]
We plan to support signed extensions for both initrd and cmdline, so you could use that. But anyway, if you have 500 initrds, you'll need a lot of space anyway. The initrds are typically bigger than kernel images, usually at least x2, and in my local Debian machine it's x10 for example. Adding this complication would just marginally decrease your disk space requirement, while making things quite a bit more convoluted.

**Comment by blitz** [comment]
Yes, this is foremost a space saving mechanism for NixOS. Because `systemd-stub` doesn't support this, we currently maintain our own [stub](https://github.com/nix-community/lanzaboote#stub) that embeds filenames + cryptographic hashes instead of the file contents itself. As much fun as it is to do Rust+UEFI programming, we would rather be using systemd-stub. :)

**Comment by poettering** [comment]
So, systemd-addons exist now, which are PE binaries that carry "add-ons" for the UKI. These PE binaries do not contain code, but simply the usual PE sections, that can be merged with or override the same sections in the UKI itself.

Right now, only the .cmdline section is hooked up for add-on support. But I guess we could hook up .initrd and .linux up too if you want?

See #27358.

This would deliver what you are asking for, but authentication of the add-ons would happen via SecureBoot/shim, not via some hash pinning inside PE.

But this should be enough for you, no?

(Patches that hook up add-ons for more PE sections are welcome)

**Comment by ElvishJerricco** [comment]
@poettering The problem with add-ons is that NixOS users need to be able to select a unique configuration from the boot menu, indexed by a NixOS generation number. So how do you deduplicate the kernel and initrd files, which are shared between generations? Keep in mind this is a many-to-many relationship. NixOS users are likely to have many generations across many kernel versions and even more initrd builds. So to deduplicate these files, we can't rely on the current add-ons design, because it will only load files that are either global or entry-specific. i.e. NixOS needs something in between, where each add-on is loaded by multiple, but not all, entries.

I think we can tolerate the *stub* being duplicated for each entry, which is the design utilized by [Lanzaboote](https://github.com/nix-community/lanzaboote/), because the stub is pretty darn small by comparison. So one option would be to add a section called `.add-ons` or something, which would list which add-ons need to be loaded and verified by SecureBoot/shim.

At the end of the day, this is basically equivalent to the original suggestion here: Add a list to the stub of which files to load and verify them somehow. It's just using SecureBoot/shim instead of a baked-in hash. Arguably, a baked-in hash is more secure, since the hash prevents the stub from being tricked into using an add-on meant for a different boot entry by a simple file rename, but I think that's a relatively minor risk. Of course, the tradeoff is that verifying with SecureBoot/shim enables the separation between vendor keys and MOKs.

**Comment by bluca** [comment]
Sorry but your approach is completely backwards. You are trying to shove a square peg in a round hole - configuring userspace via EFI boot menu or kernel configuration. It would be like wanting to select the disk decryption passphrase from the boot menu. I understand that's how grub used to do things, but there's no good reason for it, and it's in fact a misdesign.

To configure userspace, use userspace tools: credentials. Credentials are picked from the ESP and given to the initrd, so you can do your selection in your initrd userspace code. And they are just tiny text or binary files, so they can be as small as they you can make them, solving your disk space problem.

It's also going to be much easier to provide a menu in userspace, because it's userspace - UEFI mode sucks for this. As I mentioned to somebody on IRC some time ago, it might be acceptable to allow selecting which credentials to pass through among the available ones from sd-boot, but that's probably going to be _a lot_ of work compared to simply building a userspace selector that runs in the initrd. Should be easy to integrate into Plymouth for example, for a nice user experience.

But this whole idea of making UKIs hard-dependent on specific sidecars is completely opposite of what UKIs are meant to be. It should always be possible to pick a UKI, boot it, and it should just work. Addons and extensions are called addons and extensions because they are optional.

**Comment by ElvishJerricco** [comment]
@bluca NixOS users expect the boot loader menu to be how they select the generation that want to boot, and for good reason. It represents the chronological order that generations were installed in, which is *not at all tied to the order of kernel version numbers or the files present in initrd*. NixOS users upgrade and downgrade kernel versions all the time, and the expectation is that it's the generation number and the boot entry's title that inform them what they're booting. By the time you're in intrd, it's *way too late* to display this information and make the choice.

**Comment by RaitoBezarius** [comment]
> We plan to support signed extensions for both initrd and cmdline, so you could use that. But anyway, if you have 500 initrds, you'll need a lot of space anyway. The initrds are typically bigger than kernel images, usually at least x2, and in my local Debian machine it's x10 for example. Adding this complication would just marginally decrease your disk space requirement, while making things quite a bit more convoluted.

Yes, but you can have thousands of generations in NixOS and only need 15 initrds, weighing 20MB each, for a total of 300MB (and their corresponding kernels, weighing ~10MB each for an extra 70MB).

Extreme situations can fit in a reasonable 1GiB ESP.

**Comment by ElvishJerricco** [comment]
> There is no law of physics that makes that the only possible design choice

Chronological ordering of generations in the boot menu is a key feature of NixOS. It wasn't some compromise due to how grub works. The ability to rollback an upgrade by simply going up a line in the boot menu, regardless of how complex the changes to the kernel+initrd were, is a selling point. The ability to iterate on this, successively trying older and older generations is also valued. It's this chronological ordering that matters. Being presented with boot entries that only tell you the kernel version, having to guess which kernel version has the generation you want to boot, and then having to go through a second interactive menu to select the generation is not an acceptable UX for NixOS. And it only gets worse when you consider that different generations can have different initrds for the same exact kernel version. Or that there can even be multiple builds of the same version of the kernel. What a NixOS user cares about during the boot menu is first and foremost the generation number, so it should be possible to place that front and center.

**Comment by RaitoBezarius** [comment]
> There is no law of physics that makes that the only possible design choice, it's just the one you made because of how grub worked. You can change that and I'm sure the world will not end. The reason kernel+initd are separate from the rootfs is that you can set up the rootfs in a variety of complex and diverse ways. Binding sidecards together with UKIs is again square peg in round hole. Trying to make sd-boot into grub 3.0 is not going to work, and if you want to use sd-boot you need to rethink the approach.

I mean, it's fair that you are not interested into binding sidecars with UKIs. But I'm not exactly sure we can do another design decision here.

The whole idea of the NixOS generations is they act for a rootfs/filesystem-agnostic snapshot mechanism of your system configuration.
To support this, it means you need to offer ways to reconfigure at the system level, it also means, you need to be able to reconfigure the kernel and the initrd.

To the best of my knowledge, I'm not sure we can reconfigure the kernel and the initrd once we have booted in the initrd.

Assuming we do the kernel version choice thing in sd-boot, we still need to deduplicate the kernels. Now, you can say this is a design decision to let people have so many kernels, etc. This is one we cannot really change I believe.

Then, we boot into a kernel + a "thin initrd" (which should be _very_ small.) which will be in charge of loading the others initrds, so we are just reinventing a stub or systemd extension images but they have their own semantics with global/local credentials, etc, etc. It's unclear what is the cost to invest into those solutions and to make them work.

This way, we can have a proper UKI again. Though, I am still unconvinced whether systemd extension images are flexible enough for our usecases and sure law of physics says you can expand all the efforts you want to make it work but usually you prefer a shorter reasonable path if possible.

I don't think we are making sd-boot into GRUB 3.0 by doing this and we are not interested into having GRUB 3.0 anyway.

**Comment by bluca** [comment]
> > There is no law of physics that makes that the only possible design choice
> 
> Chronological ordering of generations in the boot menu is a key feature of NixOS. It wasn't some compromise due to how grub works. The ability to rollback an upgrade by simply going up a line in the boot menu, regardless of how complex the changes to the kernel+initrd were, is a selling point. The ability to iterate on this, successively trying older and older generations is also valued. It's this chronological ordering that matters. Being presented with boot entries that only tell you the kernel version, having to guess which kernel version has the generation you want to boot, and then having to go through a second interactive menu to select the generation is not an acceptable UX for NixOS. And it only gets worse when you consider that different generations can have different initrds for the same exact kernel version. Or that there can even be multiple builds of the same version of the kernel. What a NixOS user cares about during the boot menu is first and foremost the generation number, so it should be possible to place that front and center.

The point is that the choice to have a single monolithic and indivisible ID is just that, a choice that you have made. You could have separate IDs for UKI and rootfs, or you could have one ID where the first half identifies the UKI and the second half identifies the rootfs, or any other number of alternative schemes. UKI means Unified Kernel Image, if you have a separate kernel, and a separate initrd, and a separate command line, and so on, then it's no longer unified, it's back to type 1. Again, square peg in round hole. What's the point of it all, then? Why not just stick to type 1 entries, if that's what you need to meet your requirements? If you want to compromise on security and reliability for presentation, that's a choice you are allowed to make of course.

**Comment by RaitoBezarius** [comment]
> > > There is no law of physics that makes that the only possible design choice
> > 
> > 
> > Chronological ordering of generations in the boot menu is a key feature of NixOS. It wasn't some compromise due to how grub works. The ability to rollback an upgrade by simply going up a line in the boot menu, regardless of how complex the changes to the kernel+initrd were, is a selling point. The ability to iterate on this, successively trying older and older generations is also valued. It's this chronological ordering that matters. Being presented with boot entries that only tell you the kernel version, having to guess which kernel version has the generation you want to boot, and then having to go through a second interactive menu to select the generation is not an acceptable UX for NixOS. And it only gets worse when you consider that different generations can have different initrds for the same exact kernel version. Or that there can even be multiple builds of the same version of the kernel. What a NixOS user cares about during the boot menu is first and foremost the generation number, so it should be possible to place that front and center.
> 
> The point is that the choice to have a single monolithic and indivisible ID is just that, a choice that you have made. You could have separate IDs for UKI and rootfs, or you could have one ID where the first half identifies the UKI and the second half identifies the rootfs, or any other number of alternative schemes. UKI means Unified Kernel Image, if you have a separate kernel, and a separate initrd, and a separate command line, and so on, then it's no longer unified, it's back to type 1. Again, square peg in round hole. What's the point of it all, then? Why not just stick to type 1 entries, if that's what you need to meet your requirements? If you want to compromise on security and reliability for presentation, that's a choice you are allowed to make of course.

I don't think you are proving we are compromising on security and reliability, and it sounds a bit FUD to me. 

Nevertheless, you have made your point loud and clear, and that's reasonable.
We will explore how to have "proper UKIs" while not compromising on the properties we need and are interested.

Ultimately, what I gather from this exchange, is that sd-boot (and potentially UAPI group?) are not interested in supporting this design choice — I am interested though if anything with the same features will happen someday though with another design choice, hopefully, we will be able to adopt it, by then. :)

**Comment by bluca** [comment]
> I don't think you are proving we are compromising on security and reliability, and it sounds a bit FUD to me.

Maybe it was a bit unclear, I was referring specifically to type 1 entries. They are unprotected plain text files on the ESP, so the security posture is vastly worse than UKiI.

**Comment by ElvishJerricco** [comment]
@poettering I think that would be a fine solution. It has the drawback that a generation now *needs* to be booted by systemd-boot and can't be booted by UEFI directly, but I don't think that's a high priority for NixOS. @RaitoBezarius what do you think?

**Comment by RaitoBezarius** [comment]
> @poettering I think that would be a fine solution. It has the drawback that a generation now _needs_ to be booted by systemd-boot and can't be booted by UEFI directly, but I don't think that's a high priority for NixOS. @RaitoBezarius what do you think?

I mean, as we already have our own `bootctl install` procedure somewhat, we could even materialize a "self-contained" stub to boot it via UEFI directly, I imagine?

But it's indeed not high priority, I think saying "this is only supported via UKI" is completely a fair thing for now.

**Comment by poettering** [comment]
> @poettering I think that would be a fine solution. It has the drawback that a generation now _needs_ to be booted by systemd-boot and can't be booted by UEFI directly, but I don't think that's a high priority for NixOS. @RaitoBezarius what do you think?

So another TODO list item was to optionally allow sd-boot to contain one or more type1 entries embedded into a PE section (maybe in a cpio or so), that it will include in the menu. That way you could basically use sd-boot as a stub of a kind, that doesn#t even really act as a menu but just executes a type1 entry you embed.

**Comment by poettering** [comment]
If that solution works for you I guess this is what needs to be worked on:

* support for the "add-on" stanza in type1 drop-ins, including a definition how sd-stub can acquire the add-ons via a uefi protocol that sd-boot implements
* make sure "bootctl cleanup" follows the "add-on" stanza in its gc logic, too
* support loading .initrd from add-on in sd-stub


Expecting your PRs ;-)

**Comment by RaitoBezarius** [comment]
It's on our todolist! Thank you for this solution!

**Comment by RaitoBezarius** [comment]
@poettering I resumed working on this, is there a place where to discuss systemd development regarding design choices for incoming PRs?

**Comment by poettering** [comment]
@RaitoBezarius there's a certain chat room where some nixos colleagues of yours can invite you into. hey @arianvp can you invite your teammate?

but usually filing a PR is a great way to get the discussion started.

**Comment by RaitoBezarius** [comment]
> @RaitoBezarius there's a certain chat room where some nixos colleagues of yours can invite you into. hey @arianvp can you invite your teammate?
> 
> but usually filing a PR is a great way to get the discussion started.

Filled the PR, in the meantime. Thank you!

**Comment by poettering** [comment]
Small update:

PR #36314 adds a new "uki" stanza to type 1 bls, which is very similar to "linux", but means that the target EFI binary is definitely a "uki", so that we can take benefit of that.

And we now support initrd in PE addons, too.

so for your usecase you could generate type 1 bls for each of your version that points to the same uki, via the new "uki" stanza. 

There's one thing missing from our side to make you happy I guess. an ability to attach PE addons to type 1 bls. I initially thought we should add an "addon" stanza to type1 bls for that. I changed my mind on that. I think it would be nicer to simply support foo.conf.extra.d/ next a type1 bls "foo.conf" snippet, where all kinds of sidecars can be placed. This sounds a lot more powerful and retains the drop-in character of addons much better.

Implementation-wise this should probably be enabled like this: when sd-boot invokes a type 1 bls "uki" entry, check if there is a side-car directory. if so, install it as a "handle" under a new, well-known uuid. then, in sd-stub look for that handle (or in fact, allow any number of them), and then read side-cars from all such places. 

This should be very simple to add actually, and very powerfull as sd-boot and suchlike can add any kind of dir there they like, and sd-stub will happily just read the sidecars from there.

A patch for that should be less than 50 lines is my educated guess.

**Comment by RaitoBezarius** [comment]
@poettering I am watching the ongoing developments and if I find some time among my pile of stuff, I will try to play with it someday and report back :).

**Comment by arianvp** [comment]
@blitz  Yes indeed the addon code checks that the `.uname` of the addon matches the `.uname` of the UKI . This should be enough to avoid mixing  kernels and addons (and thus initrds and cmdlines)

I  guess we don't have to per se put the output of `uname -r` in the `.uname` section. We could put the full nix store path of the kernel binary or the hash of the kernel binary in it as well if we want something more unique than just the raw kernel version numbers (e.g. if someone installs new kernel with a patch). 

We can then have an addon for the cmdline and an addon for the initrd. Both with the unique fingerprint in `.uname` section to avoid mixing them with other kernels.

I think all the puzzle pieces (Except for extending BLS #1 entries to support this) are here now.

**Comment by arianvp** [comment]
@poettering just so I'm understand what you're proposing here. when you say "sidecar" you mean 

```
ESP/.../foo.conf.extra.d/*.addon.efi
ESP/.../foo.conf.extra.d/*.cred
ESP/.../foo.conf.extra.d/*.sysext.raw
ESP/.../foo.conf.extra.d/*.confext.raw
```

just like we have for UKIs ?

I suppose these only would be loaded if `foo.conf` has a `uki` option set right?

Or do we also want to support  `*.cred` , `*.{sysext,confext}.raw`  for normal  BLS#1 entries with a `kernel` and `initrd` ? I don't think so right? As I  do not think we supported `/loader/credentials/*.cred` and `/loader/extensions/*.{sysext,confext}.raw` for those so far.

**Comment by poettering** [comment]
yes, that's what i meant, and no, only for uki and uki-url, not for kernel/initrd/efi
